tools/base.sh

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env bash
+
+## Commands defined by this file:
```

Comment:
The purpose here is to avoid the prior pervasive problem of the scripts doing things almost-correctly or subtly-wrong-in-ways-that-happen-to-work-95%-of-the-time. While the current team seems good with Bash, it is apparent that past users have not been so fortunate, and future users will likely face issues as well.

I waffled back/forth on whether I really wanted to define something as simple as `is-dir` and `is-file`, but settled on it to 1) Avoid temptation to use `[` or `test`, since those are exceptionally slow on Cygwin (because they spawn subprocesses, where `[[` does not) and 2) on-site readability for less-familiar users compared to `[[ -d filepath ]]`. (I almost included a `not` command that just ran `!`, but I think people can understand `!` at a glance.)

For some things such as `run-chdir`, `have-command`, and `is-set`, they encode the idioms "as-correctly-as-possible", rather than hoping that all future script maintainers will always type the correct magic punctuation for "Does this command exist?" and "Move into a directory and run a command."